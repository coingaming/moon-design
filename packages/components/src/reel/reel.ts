import styled, { CSSObject } from 'styled-components';
import { border, colors } from '@heathmont/sportsbet-tokens';
import { spacing } from '@heathmont/sportsbet-utils';

import { listItemStyleTypeNone } from '../lists/lists';

export type ReelProps = {
  height?: CSSObject['height'];
  space?: CSSObject['margin'];
};

/**
 * Reel
 *
 * Based on the Reel component from every-layout.dev
 */
export const Reel = styled.div<ReelProps>(
  {
    display: 'flex',
    overflowX: 'auto',
    overflowY: 'hidden',
    /**
     * Scroll Behavior
     * 1. Hide Scrollbars on browsers that don't support custom scrollbars.
     * 2. Auto-hide scrollbars on IE/Edge.
     * 3. Create 'padding' around the scrollbar.
     */
    WebkitOverflowScrolling: 'touch',
    scrollbarWidth: 'none' /* [1] */,
    '-ms-overflow-style': '-ms-autohiding-scrollbar' /* [2] */,
    '::-webkit-scrollbar': {
      width: 12,
      height: 12,
      cursor: 'pointer',
    },
    '::-webkit-scrollbar-thumb': {
      backgroundColor: 'transparent',
      backgroundClip: 'content-box' /* [3] */,
      borderRadius: border.radius.largest,
      border: '3px solid transparent' /* [3] */,
    },
    ':hover::-webkit-scrollbar-thumb': {
      backgroundColor: colors.neutral[40],
    },
    /**
     * Elements
     * 1. Ensure all list-items render without bullets
     * 2. Set content width to auto
     * 3. When content is set to images, ensure they all remain the same height.
     */
    '& > li': listItemStyleTypeNone /* [1] */,
    '& > *': { flex: '0 0 auto' } /* [2] */,
    '& > img': { height: '100%', flexBasis: 'auto', width: 'auto' } /* [3] */,
  },
  ({ height = 'auto', space = spacing() }) => ({
    height,
    '& > * + *': {
      marginLeft: space,
    },
  })
);
