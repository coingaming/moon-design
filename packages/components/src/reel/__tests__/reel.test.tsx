import React from 'react';
import { create } from 'react-test-renderer';
import 'jest-styled-components';
import { sportsbetDark, ThemeProvider } from '@heathmont/moon-themes';

import Reel from '../Reel';

const renderWithTheme = (component: JSX.Element) => (
  <ThemeProvider theme={sportsbetDark}>{component}</ThemeProvider>
);

// @ts-ignore
const items = Array.from({ length: 10 }, (v, i) => i);

describe('Reel', () => {
  test('renders with defaults', () => {
    const reel = create(
      renderWithTheme(
        <Reel>
          {items.map((item) => (
            <div key={item}>Item {item}</div>
          ))}
        </Reel>,
      ),
    );

    expect(reel).toMatchSnapshot();
  });

  test('renders with custom space', () => {
    const reel = create(
      renderWithTheme(
        <Reel space={30}>
          {items.map((item) => (
            <div key={item}>Item {item}</div>
          ))}
        </Reel>,
      ),
    );

    expect(reel).toMatchSnapshot();
  });

  test('renders child images correctly with custom height', () => {
    const reel = create(
      renderWithTheme(
        <Reel height="10vh">
          {items.map((item) => (
            <img key={item} alt={`Test ${item}`} src="" />
          ))}
        </Reel>,
      ),
    );

    expect(reel).toMatchSnapshot();
  });

  test('renders child list items correctly', () => {
    const reel = create(
      renderWithTheme(
        <Reel as="ul">
          {items.map((item) => (
            <li key={item}>Item {item}</li>
          ))}
        </Reel>,
      ),
    );

    expect(reel).toMatchSnapshot();
  });
});
