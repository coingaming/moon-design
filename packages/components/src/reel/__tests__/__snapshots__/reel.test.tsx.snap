// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reel renders child images correctly with custom height 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow-x: auto;
  overflow-y: hidden;
  -webkit-overflow-scrolling: touch;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
  -ms-overflow-style: -ms-autohiding-scrollbar;
  height: 10vh;
}

.c0::-webkit-scrollbar {
  width: 12px;
  height: 12px;
  cursor: pointer;
}

.c0::-webkit-scrollbar-thumb {
  background-color: transparent;
  background-clip: content-box;
  border-radius: 100px;
  border: 3px solid transparent;
}

.c0:hover::-webkit-scrollbar-thumb {
  background-color: #62676D;
}

.c0 > li {
  list-style-type: none;
}

.c0 > li:before {
  position: absolute;
  content: "\\200B";
}

.c0 > * {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c0 > img {
  height: 100%;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  width: auto;
}

.c0 > * + * {
  margin-left: 1rem;
}

<div
  className="c0"
  height="10vh"
>
  <img
    alt="Test 0"
    src=""
  />
  <img
    alt="Test 1"
    src=""
  />
  <img
    alt="Test 2"
    src=""
  />
  <img
    alt="Test 3"
    src=""
  />
  <img
    alt="Test 4"
    src=""
  />
  <img
    alt="Test 5"
    src=""
  />
  <img
    alt="Test 6"
    src=""
  />
  <img
    alt="Test 7"
    src=""
  />
  <img
    alt="Test 8"
    src=""
  />
  <img
    alt="Test 9"
    src=""
  />
</div>
`;

exports[`Reel renders child list items correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow-x: auto;
  overflow-y: hidden;
  -webkit-overflow-scrolling: touch;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
  -ms-overflow-style: -ms-autohiding-scrollbar;
  height: auto;
}

.c0::-webkit-scrollbar {
  width: 12px;
  height: 12px;
  cursor: pointer;
}

.c0::-webkit-scrollbar-thumb {
  background-color: transparent;
  background-clip: content-box;
  border-radius: 100px;
  border: 3px solid transparent;
}

.c0:hover::-webkit-scrollbar-thumb {
  background-color: #62676D;
}

.c0 > li {
  list-style-type: none;
}

.c0 > li:before {
  position: absolute;
  content: "\\200B";
}

.c0 > * {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c0 > img {
  height: 100%;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  width: auto;
}

.c0 > * + * {
  margin-left: 1rem;
}

<ul
  className="c0"
>
  <li>
    Item 
    0
  </li>
  <li>
    Item 
    1
  </li>
  <li>
    Item 
    2
  </li>
  <li>
    Item 
    3
  </li>
  <li>
    Item 
    4
  </li>
  <li>
    Item 
    5
  </li>
  <li>
    Item 
    6
  </li>
  <li>
    Item 
    7
  </li>
  <li>
    Item 
    8
  </li>
  <li>
    Item 
    9
  </li>
</ul>
`;

exports[`Reel renders with custom space 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow-x: auto;
  overflow-y: hidden;
  -webkit-overflow-scrolling: touch;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
  -ms-overflow-style: -ms-autohiding-scrollbar;
  height: auto;
}

.c0::-webkit-scrollbar {
  width: 12px;
  height: 12px;
  cursor: pointer;
}

.c0::-webkit-scrollbar-thumb {
  background-color: transparent;
  background-clip: content-box;
  border-radius: 100px;
  border: 3px solid transparent;
}

.c0:hover::-webkit-scrollbar-thumb {
  background-color: #62676D;
}

.c0 > li {
  list-style-type: none;
}

.c0 > li:before {
  position: absolute;
  content: "\\200B";
}

.c0 > * {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c0 > img {
  height: 100%;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  width: auto;
}

.c0 > * + * {
  margin-left: 30px;
}

<div
  className="c0"
>
  <div>
    Item 
    0
  </div>
  <div>
    Item 
    1
  </div>
  <div>
    Item 
    2
  </div>
  <div>
    Item 
    3
  </div>
  <div>
    Item 
    4
  </div>
  <div>
    Item 
    5
  </div>
  <div>
    Item 
    6
  </div>
  <div>
    Item 
    7
  </div>
  <div>
    Item 
    8
  </div>
  <div>
    Item 
    9
  </div>
</div>
`;

exports[`Reel renders with defaults 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow-x: auto;
  overflow-y: hidden;
  -webkit-overflow-scrolling: touch;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
  -ms-overflow-style: -ms-autohiding-scrollbar;
  height: auto;
}

.c0::-webkit-scrollbar {
  width: 12px;
  height: 12px;
  cursor: pointer;
}

.c0::-webkit-scrollbar-thumb {
  background-color: transparent;
  background-clip: content-box;
  border-radius: 100px;
  border: 3px solid transparent;
}

.c0:hover::-webkit-scrollbar-thumb {
  background-color: #62676D;
}

.c0 > li {
  list-style-type: none;
}

.c0 > li:before {
  position: absolute;
  content: "\\200B";
}

.c0 > * {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c0 > img {
  height: 100%;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  width: auto;
}

.c0 > * + * {
  margin-left: 1rem;
}

<div
  className="c0"
>
  <div>
    Item 
    0
  </div>
  <div>
    Item 
    1
  </div>
  <div>
    Item 
    2
  </div>
  <div>
    Item 
    3
  </div>
  <div>
    Item 
    4
  </div>
  <div>
    Item 
    5
  </div>
  <div>
    Item 
    6
  </div>
  <div>
    Item 
    7
  </div>
  <div>
    Item 
    8
  </div>
  <div>
    Item 
    9
  </div>
</div>
`;
